File: 0056-merge-intervals/0056-merge-intervals.js
/**
 * @param {number[][]} intervals
 * @return {number[][]}
 */
var merge = function(intervals) {
    if (intervals.length < 2) return intervals;
    intervals.sort((a, b) => a[0] - b[0]);
    var merged = [];
    var start = intervals[0][0];
    var end = intervals[0][1];
    for (var i = 1; i < intervals.length; i++) {
        const interval = intervals[i];
        // overlap - adjust end
        if (interval[0] <= end) {
            end = Math.max(end, interval[1]);
        } else { // no overlap - add first interval
            merged.push([start, end]);
            start = interval[0];
            end = interval[1]
        }
    }
    merged.push([start, end])
    return merged;
};

File: 0056-merge-intervals/0056-merge-intervals.py
class Solution:
    def merge(self, intervals: List[List[int]]) -> List[List[int]]:
        if len(intervals) < 2:
            return intervals
        intervals.sort(key=lambda x: x[0])
        merged = []
        i = 0
        j = 1
        while (j < len(intervals)):
            if intervals[i][1] >= intervals[j][0]:
                intervals[j][0] = min(intervals[j][0], intervals[i][0])
                intervals[j][1] = max(intervals[j][1], intervals[i][1])
            else:
                merged.append(intervals[i])
            i += 1
            j += 1
        merged.append(intervals[i])
        return merged